cli/src/commands/bookmark/list.rs

```diff
@@ -270,3 +291,470 @@ struct RefListItem {
     /// Remote bookmarks tracked by the primary (or local) bookmark.
     tracked: Vec<Rc<CommitRef>>,
 }
+
+/// Sort key for the `--sort` argument option.
+#[derive(Copy, Clone, PartialEq, Debug, ValueEnum)]
+enum SortKey {
+    Name,
+    #[value(name = "name-")]
+    NameDesc,
+    AuthorName,
+    #[value(name = "author-name-")]
+    AuthorNameDesc,
+    AuthorEmail,
+    #[value(name = "author-email-")]
+    AuthorEmailDesc,
+    AuthorDate,
+    #[value(name = "author-date-")]
+    AuthorDateDesc,
+    CommitterName,
+    #[value(name = "committer-name-")]
+    CommitterNameDesc,
+    CommitterEmail,
+    #[value(name = "committer-email-")]
+    CommitterEmailDesc,
+    CommitterDate,
+    #[value(name = "committer-date-")]
+    CommitterDateDesc,
+}
+
+fn commit_for_ref_list_item(
+    item: &RefListItem,
+    f: impl Fn(CommitId) -> Option<Arc<backend::Commit>>,
+) -> Option<Arc<backend::Commit>> {
+    item.primary
+        .target()
+        .added_ids()
+        .next()
+        .cloned()
+        .and_then(f)
+}
+
+fn sort(
+    bookmark_items: &mut [RefListItem],
+    sort_keys: &[SortKey],
+    f: impl Fn(CommitId) -> Option<Arc<backend::Commit>>,
+) {
+    // Multi-pass sorting, the first key is most significant.
```

Comment:
Since we need to propagate error, maybe we can simply build an external table prior to sorting?

```rust
let commits: HashMap<CommitId, Arc<backend::Commit>> = HashMap::new();
if sort_keys.iter().any(...) { // if key depends on commit
    commits = bookmark_items...try_collect()?;
}
let to_commit = |item: &RefListItem| {
    let id = item.primary.target().added_ids().next()?;
    commits.get(id)
};
```

It will also help remove `.clone()` from `.sort_by_key()`.